src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -4312,6 +4323,21 @@ private bool HasTopLevelNullabilityConversion(TypeWithAnnotations source, TypeWi
             return operandType;
         }
 
+        private void VisitOperandsHonoringAnnotations(MethodSymbol method, ImmutableArray<BoundExpression> operands)
+        {
+            Debug.Assert(!(method is null));
+            Debug.Assert(!operands.IsDefaultOrEmpty);
+
+            ImmutableArray<FlowAnalysisAnnotations> annotations = GetAnnotations(operands.Length, expanded: false, method.Parameters, argsToParamsOpt: default);
+            if (!annotations.IsDefault)
+            {
+                bool saveDisableDiagnostics = _disableDiagnostics;
+                _disableDiagnostics = true;
+                VisitArgumentsEvaluateHonoringAnnotations(operands, method.ParameterRefKinds, annotations);
```

Comment:
>VisitArgumentsEvaluateHonoringAnnotations [](start = 16, length = 41)

looking at the other place where we use this method, we do some gymnastics to save/restore the state to avoid side-effects from double-visit. It feels likely that a  similar scenario could affect user-defined operators.

A scenario that may hit that (should have no warnings on `s.ToString()`:
```
string? s= "";
_ = s.ToString() + (s= null); // with a custom `+`
```

Reply:
I believe the state saving in the other case is because we do some other visits which can affect it, not due to `VisitArgumentsEvaluateHonoringAnnotations`

---
In reply to: [270161114](https://github.com/dotnet/roslyn/pull/34556#discussion_r270161114) [](ancestors = 270161114)